### ISSUE TITLE
No error or warning when trying to overwrite run when `.wait = FALSE`

### ISSUE THREAD

**Issue by kylebaron** [metrumresearchgroup/bbr#691]
Is this expected / documented? The simulation run was already executed. When I try to resubmit with `.wait = FALSE`, I don't get a warning or error back. If I set `.wait = TRUE`,  then I get the error. 

```r
>   submit_model(sim_mod, .wait = FALSE, .mode = "local")
Running:
  /data/home/kyleb/project.mrg/current/bin/bbi nonmem run local /data/home/kyleb/project.mrg/current/models/pk/106-sim.ctl --threads=4 --parallel
In /data/home/kyleb/project.mrg/current/models/pk
Not waiting for process to finish.
```

```r
>   submit_model(sim_mod, .wait = TRUE, .mode = "local")
Error in check_status_code(p$get_exit_status(), output, .cmd_args) :
The target output directory already exists. Please pass `.bbi_args=list(overwrite=TRUE)` to your `submit_model()` call. You can also set `overwrite: true` in the model .yaml file or the `bbi.yaml` file.
```

**Comment by kyleam** [comment]
I think it's expected in the sense that when `.wait = FALSE` we don't wait for the `bbi` subprocess to exit, and `bbi` is the one that errors when `--overwrite` isn't passed but the model directory exists.  I don't know if it's spelled out anywhere in the docs that you won't see this and other `bbi` errors when `.wait = FALSE` (the same applies to most `bbi` errors when `.mode = "sge"` even when `.wait = TRUE`).

The main options I can think of:

 * leave it be

 * document this case, if it's not already

    * Also perhaps add a helper to make it easier to get inspect the `bbi` output (which the caller can get to by reading `res$process$get_output_file()`)

 * make `bbr` check whether the output directory exists and overwrite isn't specified in any way (including through `bbi.yaml` default) as a special/common case and give a more useful error

 * adjust `bbi_exec()` to wait a short bit (say 10 ms) even when `.wait = TRUE` to see if there are any quick errors from `bbi`, and then do its standard `check_status_code` routine to propagate the error

My initial preference would be the second option.  Thoughts?

**Comment by kyleam** [comment]
I think you're probably right that users set `overwrite: true` in `bbi.yaml` much less often than setting it via other methods, but for those that do use `bbi.yaml`...

> And because we're not making it any _worse_ for the user, and are definitely making the other cases better.

... I think they'd find it worse.  They were setting `overwrite: true` in `bbi.yaml` and `bbr::submit_model` used to do as they wanted.  Now it would abort instead of overwriting.

This guard would probably be added somewhere in `submit_nonmem_model()`:

https://github.com/metrumresearchgroup/bbr/blob/88deecd4fb78b927cabf6931997e4ae157b18323/R/submit-model.R#L118-L142

We already have the `bbi.yaml` path at that point, so if we don't see overwrite in effect due to the `.overwrite` argument or in the merged `.bbi_args`, I think we should at least make some effort (more on that below) to match the `overwrite` setting bbi would see.  For some quick testing I did

```diff
diff --git a/R/submit-model.R b/R/submit-model.R
index a9c04256..ae855311 100644
--- a/R/submit-model.R
+++ b/R/submit-model.R
@@ -127,13 +127,19 @@ submit_nonmem_model <- function(.mod,

   # define working directory
   model_dir <- get_model_working_directory(.mod)
-
-  .path_exists <- file_exists(.config_path %||% file.path(model_dir, "bbi.yaml"))
-  if(!.path_exists){
+  cpath <- .config_path %||% file.path(model_dir, "bbi.yaml")
+  if(!file_exists(cpath)){
     stop(paste("No bbi configuration was found in the execution directory.",
                "Please run `bbi_init()` with the appropriate directory to continue."))
   }

+  # TODO: Explain cases where there can be a mismatch.
+  bbi_will_overwrite <- isTRUE(.bbi_args[["overwrite"]]) ||
+    isTRUE(yaml::read_yaml(cpath)[["overwrite"]])
+  if (!bbi_will_overwrite && file_exists(get_output_dir(.mod, .check_exists = FALSE))) {
+    stop("boo")
+  }
+
   if (!is.null(.config_path)) {
     cmd_args <- c(
       cmd_args,
```

I _think_ that does an okay job of matching bbi's behavior.  There are some cases it wouldn't match for due to some differences in yaml parsing (e.g., I think any non-zero integer in bbi.yaml would be treated as `true` by bbi).  Also bbi is okay with an uppercase `Overwrite`.  But it's probably good enough.

Also, as I was playing around with that, I was a bit surprised that we can't override an `overwrite: true` from `bbi.yaml` by passing `.overwrite = FALSE` (or the equivalent via `bbi_args` or the model yaml).  That makes sense, given that `.overwrite = FALSE` maps to "no argument" (rather than say `--no-overwrite`), so there is no way to distinguish between unspecified and `.overwrite = FALSE`.

**Comment by seth127** [comment]
This is still a work-in-progress, but my first shot at implementing this is in https://github.com/metrumresearchgroup/bbr/pull/692/commits/742e75f364a7a4b250126df744bc9d61bf016f65

Incidentally, I think it also addresses this point from @kyleam :

> I was a bit surprised that we can't override an overwrite: true from bbi.yaml by passing .overwrite = FALSE (or the equivalent via bbi_args or the model yaml)

Because we check the args first and then would error in the case where they are explicitly `FALSE` and a directory exists. This would essentially stop things before it even got to `bbi` to trigger the "no argument" issue. (To be clear, I don't think solving that case is _must have_ for this implementation, but if I'm not confused, I think we covered it anyway.)

**Comment by kyleam** [comment]
> Incidentally, I think it also addresses this point from @kyleam :

Yeah, it looks like you're patch checks NULL (versus TRUE) so it does stop in that case.  That may be the way to go, though I'm not sure since it now makes the handling inconsistent with all other logical values in `.bbi_args`.

**Comment by seth127** [comment]
> That may be the way to go, though I'm not sure since it now makes the handling inconsistent with all other logical values in .bbi_args.

I guess I don't have a strong opinion on this. Per your original point, I think most users would find it "surprising" that "there is no way to distinguish between unspecified and `.overwrite = FALSE`." I also see this as somewhat of a special case, in that we have an explicit `bbr` argument for `.overwrite` too, and for _that_ we [do distinguish between unspecified and `FALSE`](https://github.com/metrumresearchgroup/bbr/blob/main/R/submit-model.R#L119-L122) (or at least, that was our intention).

That said, if you think about this and feel strongly that we should keep it consistent with other `.bbi_args`, I'm fine with making that change.

**Comment by kyleam** [comment]
> also see this as somewhat of a special case, in that we have an explicit `bbr` argument for `.overwrite` too [...]

Yeah.

> That said, if you think about this and feel strongly that we should keep it consistent with other `.bbi_args`, I'm fine with making that change.

I'd say go with your change.  The inconsistency in handling compared to other bbi flags is making me pause (as well as the desire to make the minimal number of behavioral changes), but I bet the change leads to less confusion than more overall.

**Comment by barrettk** [comment]
This was addressed for the `submit_model` methods in https://github.com/metrumresearchgroup/bbr/pull/692, however `submit_models()` (plural) is still an open question; mainly how we want to handle `.bbi_args()`

I suggest we keep similar handling, and perhaps check that the line `isTRUE(yaml::read_yaml(cpath)[["overwrite"]])` is `TRUE` for all models in `.mods` ([see ref](https://github.com/metrumresearchgroup/bbr/blob/patch/overwrite_presubmission/R/submit-model.R#L256)).

**Comment by seth127** [comment]
Thanks for posting that detail @barrettk . Looking at it quickly, the main issue that complicates things is that `submit_models()` allows different models to be carrying around different `bbi_args` and then reconciles the unique combinations or args into distinct `bbi` calls (code [here](https://github.com/metrumresearchgroup/bbr/blob/88deecd4fb78b927cabf6931997e4ae157b18323/R/utils.R#L139-L186), called [here](https://github.com/metrumresearchgroup/bbr/blob/67bd15633b1c7f3815300c96eadc0048a0c66710/R/submit-models.R#L105)).

Without giving it much more thought, I'm not sure how we want to handle that. I think ideally we would decide on this and implement before the next release (which will have the #692 changes in it). That said, if remains as is for now, I'm still fine releasing those #692 changes, unless someone else objects.
